runtime-sdk/src/modules/consensus_accounts/mod.rs

```diff
@@ -0,0 +1,422 @@
+    fn default() -> Self {
+        Self {
+            // TODO: get this from consensus? (token symbol?).
+            consensus_denomination: token::Denomination::from_str("TEST").expect("to work"),
+        }
+    }
+}
+
+impl module::Parameters for Parameters {
+    type Error = ();
+}
+
+/// Genesis state for the consensus module.
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct Genesis {
+    #[serde(rename = "parameters")]
+    pub parameters: Parameters,
+}
+
+/// Interface that can be called from other modules.
+pub trait API {
+    /// Deposit an amount into the runtime account.
+    fn deposit<C: Context>(
+        ctx: &mut C,
+        from: Address,
+        amount: token::BaseUnits,
+    ) -> Result<(), Error>;
+
+    /// Withdraw an amount out from the runtime account.
+    fn withdraw<C: Context>(
+        ctx: &mut C,
+        to: Address,
+        amount: token::BaseUnits,
+    ) -> Result<(), Error>;
+
+    // TODO:
+    //  - Add/reclaim deposited escrow.
+    //      - need a way to get escrow events in runtime: https://github.com/oasisprotocol/oasis-core/issues/3862
+}
+
+/// Consensus accounts module.
+///
+/// This module allows consensus transfers in and out of the runtime account,
+/// while keeping track of amount deposited per account.
+pub struct Module<Accounts: modules::accounts::API, Consensus: modules::consensus::API> {
+    _accounts: std::marker::PhantomData<Accounts>,
+    _consensus: std::marker::PhantomData<Consensus>,
+}
+
+const CONSENSUS_TRANSFER_HANDLER: &str = "consensus.TransferFromRuntime";
+const CONSENSUS_WITHDRAW_HANDLER: &str = "consensus.WithdrawIntoRuntime";
+
+impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API> API
+    for Module<Accounts, Consensus>
+{
+    fn deposit<C: Context>(
+        ctx: &mut C,
+        from: Address,
+        amount: token::BaseUnits,
+    ) -> Result<(), Error> {
+        if ctx.is_check_only() {
+            return Ok(());
+        }
+
+        // Ensure correct denomination.
+        let params = Self::params(ctx.runtime_state());
+        let denomination = params.consensus_denomination;
+        if amount.denomination() != &denomination {
+            return Err(Error::InvalidDenomination);
+        }
+
+        // XXX: could check consensus state if allowance for the runtime account
+        // exists, but consensus state could be outdated since last block, so
+        // just try to withdraw.
+
+        // Do withdraw from the consensus account and update the account state if
+        // successful.
+        Consensus::withdraw(
+            ctx,
+            from,
+            &amount,
+            MessageEventHookInvocation {
+                hook_name: CONSENSUS_WITHDRAW_HANDLER.to_string(),
+                payload: cbor::to_value(types::ConsensusWithdrawContext {
+                    address: from,
+                    amount: amount.clone(),
+                }),
+            },
```

Comment:
Could add a helper (e.g. a generic method like `MessageEventHookInvocation::new(hook_name, payload)` where `payload` must implement the `Serialize` trait) for this so that you don't need to call `to_string` and `to_value` explicitly?

Reply:
Good idea, thanks!